### ISSUE TITLE
Creating too many windows-rs COM interfaces breaks IntelliJ-Rust

### ISSUE THREAD

**Issue by LoganDark** [intellij-rust/intellij-rust#9135]
## Environment

* **IntelliJ Rust plugin version:** 0.4.176.4779-222-nightly
* **Rust toolchain version:** 1.64.0-nightly (3924dac7b 2022-07-29) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2022.2 (CL-222.3345.126)
* **Operating system:** Windows 10 2004 build 19041.572 (64-bit)
* **Macro expansion engine:** new
* **Name resolution engine:** new
* **Additional experimental features:** org.rust.cargo.features.settings.gutter, org.rust.cargo.emulate.terminal, org.rust.cargo.evaluate.build.scripts, org.rust.macros.proc

## Problem description

Creating a bunch of windows-rs COM interfaces broke IntelliJ-Rust. Syntax highlighting, refactoring and autocomplete do not work. Apparently, up to a certain point, the `interface` macro works fine, with syntax highlighting and refactoring and autocomplete. Then it reaches some critical point and IntelliJ-Rust decides that you don't deserve its help anymore.

Transplanting the source files into another project breaks it in the same way. Invalidating caches, deleting .idea and target (all at once) does not fix it. Only completely removing the `mod` declaration will fix it.

## Steps to reproduce

place this module

[realworld.zip](https://github.com/intellij-rust/intellij-rust/files/9227889/realworld.zip)

in a project with this dependency

```toml
windows = { version = '~0.39.0', features = ['interface', 'implement', 'Win32_System_Com', 'Win32_Foundation', 'Win32_UI_WindowsAndMessaging'] }
```

The entire project will lose all analysis and refactoring and intellisense until the module is removed.

**Comment by afetisov** [comment]
I don't observe the problem you mentioned. I have created a new project, added the dependency as above, and added `mod realword;` to `main.rs`. The contents of `realworld` have some trouble with indexing and macro expansion, many of the symbols can't properly resolve. This is unfortunate, but also not unexpected with proc macros. However, the analyses are still performed, the items in `main.rs` which don't depend on `realworld` can be analyzed an refactored without any problem.

I think you need to provide a more specific reproducing example for your issue. What's the project where the error happens, and what specifically can't be refactored?

By the way, the system info you posted says you are using `x86_64-unknown-linux-gnu` target on Windows with the windows crate. That sounds... weird? Maybe the windows crate simply can't compile with your toolchain? I have observed that errors in dependencies can indeed cause the plugin to loose some of its functionality (e.g. all clippy lint marks), though refactoring local items should not be affected.

**Comment by LoganDark** [comment]
> I don't observe the problem you mentioned. I have created a new project, added the dependency as above, and added `mod realword;` to `main.rs`. The contents of `realworld` have some trouble with indexing and macro expansion, many of the symbols can't properly resolve. This is unfortunate, but also not unexpected with proc macros. However, the analyses are still performed, the items in `main.rs` which don't depend on `realworld` can be analyzed an refactored without any problem.
> 
> I think you need to provide a more specific reproducing example for your issue. What's the project where the error happens, and what specifically can't be refactored?
> 
> By the way, the system info you posted says you are using `x86_64-unknown-linux-gnu` target on Windows with the windows crate. That sounds... weird? Maybe the windows crate simply can't compile with your toolchain? I have observed that errors in dependencies can indeed cause the plugin to loose some of its functionality (e.g. all clippy lint marks), though refactoring local items should not be affected.

Yeah I don't know why CLion chose to use the WSL toolchain, but I switched to the Windows toolchain a while ago and while everything across the project is still painfully slow, it at least *works* if you stop typing and wait about 30 seconds for the autocomplete to show up.

It's possible that analysis and syntax highlighting still *could* happen before, but it was so painfully slow I never waited around long enough to let it complete.

**Comment by afetisov** [comment]
The macros in my example project took a while to expand, but I didn't have any issues once they have finished expanding. However, the files I edited didn't have any `interface` macro calls in them. I assume that if you are trying to edit something which is related to those interface definitions, the performance will be worse.

**Comment by LoganDark** [comment]
> The macros in my example project took a while to expand, but I didn't have any issues once they have finished expanding. However, the files I edited didn't have any `interface` macro calls in them. I assume that if you are trying to edit something which is related to those interface definitions, the performance will be worse.

I quite literally mean the entire project even outside of the realworld module has extremely bad analysis. One of the most annoying issues is that it's impossible to coerce the plugin into filling in method signatures for me in a trait impl:

https://user-images.githubusercontent.com/4723091/182355474-18af257e-d232-4e26-9b50-6528b1e57c00.mp4

(These traits were generated by the proc macro, but before a certain point, all of my interfaces had syntax highlighting and analysis perfectly fine. Now, IntelliJ-Rust is refusing to do anything about that.

If I reopen a new project from scratch which *only* defines one interface in lib.rs, this shows up for a few seconds:

![image](https://user-images.githubusercontent.com/4723091/182356281-f48ccb22-bbd2-4124-bda4-ffe5fdbfa4e5.png)

but then the plugin decides "Yeah Nah" and blows all the syntax highlighting away.

**Comment by afetisov** [comment]
Could you provide a complete project example? I really don't feel like typing "E0DAwhatever" by hand.

**Comment by LoganDark** [comment]
Sorry, the GUID itself is not important in any way so I didn't think you'd try to type it all out by hand. Here is a complete project example.

[thing.zip](https://github.com/intellij-rust/intellij-rust/files/9243701/thing.zip)

**Comment by afetisov** [comment]
I have opened your project, and was still unable to reproduce your errors. Everything works as expected for me. I can add implementations of trait methods, I can rename parameters, or change signature, or go to trait definition.

I have used the exact windows nightly toolchain version that you specified, and used the experimental features that you have mentioned. The only difference is that I have a more recent nightly build of the plugin, and a more recent build of Windows 10, but I would be surprised if that was the reason everything works.

**Comment by LoganDark** [comment]
> I have opened your project, and was still unable to reproduce your errors. Everything works as expected for me. I can add implementations of trait methods, I can rename parameters, or change signature, or go to trait definition.

If you add this code in lib.rs:

```rust


#[windows::core::implement(ILocalizedString)]
pub struct Whatever;

impl ILocalizedString_Impl for Whatever {
	
}
```

then you won't be able to fill in the trait methods for ILocalizedString_Impl

> I have used the exact windows nightly toolchain version that you specified, and used the experimental features that you have mentioned. The only difference is that I have a more recent nightly build of the plugin, and a more recent build of Windows 10, but I would be surprised if that was the reason everything works.

Yeah I installed the latest nightly plugin and it didn't fix the issue, so your assessment is correct.

> Do you use the standard installation of Rust? Are you sure you're not using WSL? It seems impossible that VS build tools could cause that bug, but just in case, are you sure they are installed correctly? Maybe you have a cargo workspace with some other crates, or maybe you have vendored some of your dependencies?

1. yes
2. ![image](https://user-images.githubusercontent.com/4723091/182496200-f7aaed85-0443-4a68-b9b2-42b66dbc3919.png)
3. VS build tools should be fine
4. No cargo workspace or dependency vendoring
